README.md
# vscode-juejin README

juejin 浏览器插件 vscode 版本


### Todo List 

- [X] 滚动加载
- [X] 样式优化
- [X] bug 处理




CHANGELOG.md
# Change Log

All notable changes to the "vscode-juejin" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/api.ts
import * as vscode from 'vscode';
import axios from 'axios';

export enum SortTypeEnum {
  hot = 200,
  new = 300
}

export interface IArticleReq {
  cate_id: string;
  client_type: number;
  cursor: string;
  id_type: number;
  limit: number;
  sort_type: SortTypeEnum;
}

export function getArticles(params: IArticleReq) {
  return axios.post('https://api.juejin.cn/recommend_api/v1/article/recommend_cate_feed?uuid=7032205944063542798&aid=6587', params).then(({ data }) => {
    return data;
  }).catch((e) => {
    vscode.window.showInformationMessage('获取数据失败');
  });
}



export enum GithubCategoryEnum {
  trending = 'trending',
  upcome = 'upcome'
}

export enum GithubPeriodEnum {
  day = 'day',
  week = 'week',
  month = 'month'
}

export enum GithubLangEnum {
  javascript = 'javascript',
  rust = 'rust',
  typescript = 'typescript',
  vue = 'vue',
  css = 'css',
  html = 'html'
}

export interface IGithubReq {
  category: GithubCategoryEnum;
  lang: GithubLangEnum;
  limit: number;
  offset: number;
  period: GithubPeriodEnum;
}

export function getGithub(params: IGithubReq) {
  return axios.post('https://e.juejin.cn/resources/github', params).then(({ data }) => data).catch((e) => {
    vscode.window.showInformationMessage('获取数据失败');
  });
}


web/config/paths.js
'use strict';

const path = require('path');
const fs = require('fs');
const getPublicUrlOrPath = require('react-dev-utils/getPublicUrlOrPath');

// Make sure any symlinks in the project folder are resolved:
// https://github.com/facebook/create-react-app/issues/637
const appDirectory = fs.realpathSync(process.cwd());
const resolveApp = relativePath => path.resolve(appDirectory, relativePath);

// We use `PUBLIC_URL` environment variable or "homepage" field to infer
// "public path" at which the app is served.
// webpack needs to know it to put the right <script> hrefs into HTML even in
// single-page apps that may serve index.html for nested URLs like /todos/42.
// We can't use a relative path in HTML because we don't want to load something
// like /todos/42/static/js/bundle.7289d.js. We have to know the root.
const publicUrlOrPath = getPublicUrlOrPath(
  process.env.NODE_ENV === 'development',
  require(resolveApp('package.json')).homepage,
  process.env.PUBLIC_URL
);


const buildPath = path.resolve(appDirectory, '../web_dist');

const moduleFileExtensions = [
  'web.mjs',
  'mjs',
  'web.js',
  'js',
  'web.ts',
  'ts',
  'web.tsx',
  'tsx',
  'json',
  'web.jsx',
  'jsx',
];

// Resolve file paths in the same order as webpack
const resolveModule = (resolveFn, filePath) => {
  const extension = moduleFileExtensions.find(extension =>
    fs.existsSync(resolveFn(`${filePath}.${extension}`))
  );

  if (extension) {
    return resolveFn(`${filePath}.${extension}`);
  }

  return resolveFn(`${filePath}.js`);
};

// config after eject: we're in ./config/
module.exports = {
  dotenv: resolveApp('.env'),
  appPath: resolveApp('.'),
  appBuild: resolveApp(buildPath),
  appPublic: resolveApp('public'),
  appHtml: resolveApp('public/index.html'),
  appIndexJs: resolveModule(resolveApp, 'src/index'),
  appPackageJson: resolveApp('package.json'),
  appSrc: resolveApp('src'),
  appTsConfig: resolveApp('tsconfig.json'),
  appJsConfig: resolveApp('jsconfig.json'),
  yarnLockFile: resolveApp('yarn.lock'),
  testsSetup: resolveModule(resolveApp, 'src/setupTests'),
  proxySetup: resolveApp('src/setupProxy.js'),
  appNodeModules: resolveApp('node_modules'),
  swSrc: resolveModule(resolveApp, 'src/service-worker'),
  publicUrlOrPath,
};



module.exports.moduleFileExtensions = moduleFileExtensions;


web/config/webpackDevServer.config.js
'use strict';

const fs = require('fs');
const errorOverlayMiddleware = require('react-dev-utils/errorOverlayMiddleware');
const evalSourceMapMiddleware = require('react-dev-utils/evalSourceMapMiddleware');
const noopServiceWorkerMiddleware = require('react-dev-utils/noopServiceWorkerMiddleware');
const ignoredFiles = require('react-dev-utils/ignoredFiles');
const redirectServedPath = require('react-dev-utils/redirectServedPathMiddleware');
const paths = require('./paths');
const getHttpsConfig = require('./getHttpsConfig');

const host = process.env.HOST || '0.0.0.0';
const sockHost = process.env.WDS_SOCKET_HOST;
const sockPath = process.env.WDS_SOCKET_PATH; // default: '/sockjs-node'
const sockPort = process.env.WDS_SOCKET_PORT;

module.exports = function (proxy, allowedHost) {
  return {
    writeToDisk: true,
    // WebpackDevServer 2.4.3 introduced a security fix that prevents remote
    // websites from potentially accessing local content through DNS rebinding:
    // https://github.com/webpack/webpack-dev-server/issues/887
    // https://medium.com/webpack/webpack-dev-server-middleware-security-issues-1489d950874a
    // However, it made several existing use cases such as development in cloud
    // environment or subdomains in development significantly more complicated:
    // https://github.com/facebook/create-react-app/issues/2271
    // https://github.com/facebook/create-react-app/issues/2233
    // While we're investigating better solutions, for now we will take a
    // compromise. Since our WDS configuration only serves files in the `public`
    // folder we won't consider accessing them a vulnerability. However, if you
    // use the `proxy` feature, it gets more dangerous because it can expose
    // remote code execution vulnerabilities in backends like Django and Rails.
    // So we will disable the host check normally, but enable it if you have
    // specified the `proxy` setting. Finally, we let you override it if you
    // really know what you're doing with a special environment variable.
    disableHostCheck:
      !proxy || process.env.DANGEROUSLY_DISABLE_HOST_CHECK === 'true',
    // Enable gzip compression of generated files.
    compress: true,
    // Silence WebpackDevServer's own logs since they're generally not useful.
    // It will still show compile warnings and errors with this setting.
    clientLogLevel: 'none',
    // By default WebpackDevServer serves physical files from current directory
    // in addition to all the virtual build products that it serves from memory.
    // This is confusing because those files won’t automatically be available in
    // production build folder unless we copy them. However, copying the whole
    // project directory is dangerous because we may expose sensitive files.
    // Instead, we establish a convention that only files in `public` directory
    // get served. Our build script will copy `public` into the `build` folder.
    // In `index.html`, you can get URL of `public` folder with %PUBLIC_URL%:
    // <link rel="icon" href="%PUBLIC_URL%/favicon.ico">
    // In JavaScript code, you can access it with `process.env.PUBLIC_URL`.
    // Note that we only recommend to use `public` folder as an escape hatch
    // for files like `favicon.ico`, `manifest.json`, and libraries that are
    // for some reason broken when imported through webpack. If you just want to
    // use an image, put it in `src` and `import` it from JavaScript instead.
    contentBase: paths.appPublic,
    contentBasePublicPath: paths.publicUrlOrPath,
    // By default files from `contentBase` will not trigger a page reload.
    watchContentBase: true,
    // Enable hot reloading server. It will provide WDS_SOCKET_PATH endpoint
    // for the WebpackDevServer client so it can learn when the files were
    // updated. The WebpackDevServer client is included as an entry point
    // in the webpack development configuration. Note that only changes
    // to CSS are currently hot reloaded. JS changes will refresh the browser.
    hot: true,
    // Use 'ws' instead of 'sockjs-node' on server since we're using native
    // websockets in `webpackHotDevClient`.
    transportMode: 'ws',
    // Prevent a WS client from getting injected as we're already including
    // `webpackHotDevClient`.
    injectClient: false,
    // Enable custom sockjs pathname for websocket connection to hot reloading server.
    // Enable custom sockjs hostname, pathname and port for websocket connection
    // to hot reloading server.
    sockHost,
    sockPath,
    sockPort,
    // It is important to tell WebpackDevServer to use the same "publicPath" path as
    // we specified in the webpack config. When homepage is '.', default to serving
    // from the root.
    // remove last slash so user can land on `/test` instead of `/test/`
    publicPath: paths.publicUrlOrPath.slice(0, -1),
    // WebpackDevServer is noisy by default so we emit custom message instead
    // by listening to the compiler events with `compiler.hooks[...].tap` calls above.
    quiet: true,
    // Reportedly, this avoids CPU overload on some systems.
    // https://github.com/facebook/create-react-app/issues/293
    // src/node_modules is not ignored to support absolute imports
    // https://github.com/facebook/create-react-app/issues/1065
    watchOptions: {
      ignored: ignoredFiles(paths.appSrc),
    },
    https: getHttpsConfig(),
    host,
    overlay: false,
    historyApiFallback: {
      // Paths with dots should still use the history fallback.
      // See https://github.com/facebook/create-react-app/issues/387.
      disableDotRule: true,
      index: paths.publicUrlOrPath,
    },
    public: allowedHost,
    // `proxy` is run between `before` and `after` `webpack-dev-server` hooks
    proxy,
    before(app, server) {
      // Keep `evalSourceMapMiddleware` and `errorOverlayMiddleware`
      // middlewares before `redirectServedPath` otherwise will not have any effect
      // This lets us fetch source contents from webpack for the error overlay
      app.use(evalSourceMapMiddleware(server));
      // This lets us open files from the runtime error overlay.
      app.use(errorOverlayMiddleware());

      if (fs.existsSync(paths.proxySetup)) {
        // This registers user provided middleware for proxy reasons
        require(paths.proxySetup)(app);
      }
    },
    after(app) {
      // Redirect to `PUBLIC_URL` or `homepage` from `package.json` if url not match
      app.use(redirectServedPath(paths.publicUrlOrPath));

      // This service worker file is effectively a 'no-op' that will reset any
      // previous service worker registered for the same host:port combination.
      // We do this in development to avoid hitting the production cache if
      // it used the same host and port.
      // https://github.com/facebook/create-react-app/issues/2272#issuecomment-302832432
      app.use(noopServiceWorkerMiddleware(paths.publicUrlOrPath));
    },
  };
};


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import {
  getArticles,
  getGithub,
  GithubCategoryEnum,
  GithubLangEnum,
  GithubPeriodEnum,
} from './api';

let cachedPanel: vscode.WebviewPanel | null;

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  let disposable = vscode.commands.registerCommand(
    'vscode-juejin.jueJin',
    () => {

      if (cachedPanel) {
        cachedPanel.reveal(vscode.ViewColumn.One);
        return;
      }
      const panel = vscode.window.createWebviewPanel(
        'article',
        '掘金插件',
        vscode.ViewColumn.One,
        {
          enableScripts: true,
          localResourceRoots: [
            vscode.Uri.joinPath(context.extensionUri, 'media'),
            vscode.Uri.joinPath(context.extensionUri, 'web_dist'),
          ],
        }
      );

      panel.webview.html = getHtmlForWebview(
        panel.webview,
        context.extensionUri
      );

      panel.onDidDispose(() => {
        panel.dispose();
        cachedPanel = null;
      });

      panel.webview.onDidReceiveMessage(
        (message) => {
          let params = null;
          try {
            params = JSON.parse(message.data);
          } catch (e) {
            return;
          }


          switch (message.type) {
            case 'fetch:articles':
              if (!params) {
                params = {
                  cate_id: '6809637767543259144',
                  client_type: 6587,
                  cursor: '0',
                  id_type: 2,
                  limit: 20,
                  sort_type: 200,
                };
              }
              getArticles(params).then((data) => {
                panel.webview.postMessage({ type: 'fetched:articles', data });
              });
              break;
            case 'fetch:githubs':
              if (!params) {
                params = {
                  category: GithubCategoryEnum.trending,
                  lang: GithubLangEnum.javascript,
                  limit: 20,
                  offset: 0,
                  period: GithubPeriodEnum.day,
                };
              }
              getGithub(params).then(({ data }) => {
                panel.webview.postMessage({ type: 'fetched:githubs', data });
              });
              break;
            default:
              console.error('Unrecognized event type');
              break;
          }
        },
        undefined,
        context.subscriptions
      );

      cachedPanel = panel;
    }
  );

  context.subscriptions.push(disposable);
}

function getHtmlForWebview(webview: vscode.Webview, extensionUri: vscode.Uri) {
  // Local path to main script run in the webview
  const scriptPathOnDisk = vscode.Uri.joinPath(
    extensionUri,
    'web_dist/js',
    'bundle.js'
  );

	const stylesMainUri = webview.asWebviewUri(vscode.Uri.joinPath(extensionUri, 'web_dist/css',
	'main.css'));

  // And the uri we use to load this script in the webview
  const scriptUri = scriptPathOnDisk.with({ scheme: 'vscode-resource' });

  // Local path to css styles
  // const styleResetPath = vscode.Uri.joinPath(extensionUri, 'media', 'reset.css');
  // const stylesPathMainPath = vscode.Uri.joinPath(extensionUri, 'media', 'vscode.css');

  // // Uri to load styles into webview
  // const stylesResetUri = webview.asWebviewUri(styleResetPath);
  // const stylesMainUri = webview.asWebviewUri(stylesPathMainPath);

  // Use a nonce to only allow specific scripts to be run
  const nonce = getNonce();

  return `<!DOCTYPE html>
		<html lang="en">
		<head>
			<meta charset="UTF-8">

			<!--
				Use a content security policy to only allow loading images from https or from our extension directory,
				and only allow scripts that have a specific nonce.
			-->

			<meta name="viewport" content="width=device-width, initial-scale=1.0">
			${process.env.NODE_ENV === 'development' ? '' : `<link href="${stylesMainUri}" rel="stylesheet">`}

			<title>掘金 vscode 插件</title>
		</head>
		<body>
			<div id="root"></div>
			<script src="${scriptUri}"></script>
		</body>
		</html>`;
}

function getNonce() {
  let text = '';
  const possible =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  for (let i = 0; i < 32; i++) {
    text += possible.charAt(Math.floor(Math.random() * possible.length));
  }
  return text;
}

// this method is called when your extension is deactivated
export function deactivate() {
  if (cachedPanel) {
    cachedPanel.dispose();
  }
}
